src/uct/ugni/smsg/ugni_smsg_iface.c

```diff
@@ -0,0 +1,456 @@
+
+    /* Yes, GNI_RC_NOT_DONE means that you're done with the smsg mailbox */
+    if(GNI_RC_NOT_DONE == ugni_rc){
+      pthread_mutex_unlock(&uct_ugni_global_lock);
+      return;
+    }
+
+    if(GNI_RC_SUCCESS != ugni_rc){
+      ucs_error("Unhandled smsg error: %s %d", gni_err_str[ugni_rc], ugni_rc);
+      pthread_mutex_unlock(&uct_ugni_global_lock);
+      return;
+    }
+
+    if(NULL == data_ptr){
+      ucs_error("Empty data pointer in smsg.");
+      pthread_mutex_unlock(&uct_ugni_global_lock);
+      return;
+    }
+
+    header = (uct_ugni_smsg_header_t *)data_ptr;
+    user_data = (void *)(header + 1);
+    void *user_desc = iface->user_desc+1;
+
+    uct_iface_trace_am(&iface->super.super, UCT_AM_TRACE_TYPE_RECV,
+                       tag, user_data, header->length, "RX: AM");
+
+    status = uct_iface_invoke_am(&iface->super.super, tag, user_data,
+                                 header->length, user_desc);
+
+    if(status != UCS_OK){
+        uct_recv_desc_iface(user_desc) = &iface->super.super.super;
+        UCT_TL_IFACE_GET_TX_DESC(&iface->super.super, &iface->free_desc,
+                                 iface->user_desc, iface->user_desc = NULL);
+    }
+
+    GNI_SmsgRelease(ep->super.ep);
+  }
+}
+
+static void uct_ugni_smsg_handle_remote_overflow(uct_ugni_smsg_iface_t *iface){
+  gni_return_t ugni_rc;
+  gni_cq_entry_t event_data;
+  struct sglib_hashed_uct_ugni_ep_t_iterator ep_iterator;
+  uct_ugni_ep_t *current_ep;
+  uct_ugni_smsg_ep_t *ep;
+
+  /* We don't know which EP dropped a completion entry, so flush everything */
+  do{
+    ugni_rc = GNI_CqGetEvent(iface->remote_cq, &event_data);
+  } while(GNI_RC_NOT_DONE != ugni_rc);
+
+  current_ep = sglib_hashed_uct_ugni_ep_t_it_init(&ep_iterator, iface->super.eps);
+  while(current_ep != NULL){
+    ep = ucs_derived_of(current_ep, uct_ugni_smsg_ep_t);
+    process_mbox(iface, ep);
+    current_ep = sglib_hashed_uct_ugni_ep_t_it_next(&ep_iterator);
+  }
+}
+
+static void progress_remote_cq(uct_ugni_smsg_iface_t *iface){
+  gni_return_t ugni_rc;
+  gni_cq_entry_t event_data;
+  uct_ugni_ep_t tl_ep;
+  uct_ugni_ep_t *ugni_ep;
+  uct_ugni_smsg_ep_t *ep;
+
+  ugni_rc = GNI_CqGetEvent(iface->remote_cq, &event_data);
+
+  if(GNI_RC_NOT_DONE == ugni_rc){
+    return;
+  }
+
+  if (GNI_RC_SUCCESS != ugni_rc || !GNI_CQ_STATUS_OK(event_data) || GNI_CQ_OVERRUN(event_data)) {
+    if(GNI_RC_ERROR_RESOURCE == ugni_rc || (GNI_RC_SUCCESS == ugni_rc && GNI_CQ_OVERRUN(event_data))){
+      uct_ugni_smsg_handle_remote_overflow(iface);
+      return;
+    }
+    ucs_error("GNI_CqGetEvent falied with unhandled error. Error status %s %d ",
+              gni_err_str[ugni_rc], ugni_rc);
+    return;
+  }
+
+  tl_ep.hash_key = GNI_CQ_GET_INST_ID(event_data);
+  ugni_ep = sglib_hashed_uct_ugni_ep_t_find_member(iface->super.eps, &tl_ep);
+  ep = ucs_derived_of(ugni_ep, uct_ugni_smsg_ep_t);
+
+  process_mbox(iface, ep);
```

Comment:
Just to be safe I would declare it as a static inline (right now it is just static)